File: leetcode/c++/medium/_2554_maximum_number_of_integers_to_choose_from_a_range_1.cpp
// Leetcode Problem 2554

#include <gtest/gtest.h>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int maxCount(vector<int>& banned , int n, int maxSum) {
        unordered_set<int> banned_numbers(banned.begin(), banned.end());

        int sum = 0;
        int amount = 0;

        for (int i = 1; i <= n; i++) {
            if (banned_numbers.contains(i)) {
                continue;
            }

            sum += i;
            if (sum > maxSum) {
                break;
            }
            amount++;
        }

        return amount;
    }
};


class TestMaxCount : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestMaxCount, SimpleTestCase) {
    vector<int> banned = {1, 6, 5};

    EXPECT_EQ(solution.maxCount(banned, 5, 6), 2);
}

TEST_F(TestMaxCount, NoIntegerIsAvailable) {
    vector<int> banned = {1, 2, 3, 4, 5, 6, 7};

    EXPECT_EQ(solution.maxCount(banned, 8, 1), 0);
}

TEST_F(TestMaxCount, SimpleTestCase2) {
    vector<int> bannend = {11};

    EXPECT_EQ(solution.maxCount(bannend, 7, 50), 7);
}

File: leetcode/rust/problems/src/medium/_2554_maximum_number_of_integers_to_choose_from_a_range_1.rs
// Leetcode Problem 2554

use std::collections::HashSet;

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn max_count(banned: Vec<i32>, n: i32, max_sum: i32) -> i32 {
        let banned_numbers: HashSet<i32> = banned.into_iter().collect();

        let mut sum = 0;
        let mut amount = 0;

        for i in 1..=n {
            if banned_numbers.contains(&i) {
                continue;
            }
            
            sum += i;
            if sum > max_sum {
                break;
            }
            amount += 1;
        }

        amount
    }
}


#[test]
fn test_max_count() {
    let banned1 = vec![1, 6, 5];
    let banned2 = vec![1, 2, 3, 4, 5, 6, 7];
    let banned3 = vec![11];

    assert_eq!(Solution::max_count(banned1, 5, 6), 2);
    assert_eq!(Solution::max_count(banned2, 8, 1), 0);
    assert_eq!(Solution::max_count(banned3, 7, 50), 7);
}